### PR TITLE
GPSd in Web UI

### PR DESCRIPTION
**Pull request by blockmurder**
Complete refactoring of #146 in a new branch. Hope I got all the changes we discussed in #146 @martinhpedersen. Ping #132

### CODE REVIEW

**Comment by martinhpedersen** [comment]
Great! I'm looking forward to reviewing it :)

One quick request: I'd like to keep the git history as clean as possible. Could you revert the last merge commit and instead rebase the branch? Just let me know if you need help, or would prefer that I do it myself.

Thanks!

**Comment by martinhpedersen** [comment]
Of course, after revert and rebase you will have to force push the changes to this branch.

**Comment by blockmurder** [comment]
Done

**Comment by martinhpedersen** [comment]
Hmm... did you rebase on top of the upstream develop branch as well? Looks like there is a conflict, but that should not be the case after the rebase.

**Comment by blockmurder** [comment]
I resolved this conflict while merging your current develop branch into feature/gpsdUi. As I have reverted this now, the conflict pops up again.

**Comment by martinhpedersen** [comment]
I see. That's what [rebase](https://git-scm.com/docs/git-rebase) are for. But no worries, I can do it once we're ready for the merge :)

**Comment by blockmurder** [comment]
Sorry, my bad. I clearly did not read your comment properly! Must be the bad sea state we currently have ;-)

**Comment by martinhpedersen** [comment]
Haha! I love the enthusiasm 😂

File: cfg/config.go [hunk 1]
```diff
@@ -56,8 +56,13 @@ type Config struct {
...
 
-	// Address and port to a GPSd daemon for position reporting.
-	GPSdAddr string `json:"gpsd_addr"`
+	// WARNING: If you enable GPSd http endpoint (enable_http) you might
+	// expose your current position to anyone who has access to Pat!!!
+	// See GPSdConfig.
+	GPSd GPSdConfig `json:gpsd`
+
+	// Legacy support for old config files
+	GPSdAddrLegacy string `json:"gpsd_addr"`
```
[the remark this answers is not in the document]
Comment (blockmurder) [reply cfg/config.go]: Oh, this is quite a nice feature, good to know!

File: http.go [hunk 2]
```diff
@@ -61,6 +63,15 @@ var websocketHub *WSHub
 func ListenAndServe(addr string) error {
 	log.Printf("Starting HTTP service (%s)...", addr)
 
+	if host, _, _ := net.SplitHostPort(addr); host == "" && config.GPSd.EnableHttp {
+		// can't use fmt.Printf as it would not show on the web interface
```
[the remark this answers is not in the document]
Comment (blockmurder) [reply http.go]: I see, I have to look up how the logger works. Good to know!

File: http.go [hunk 3]
```diff
@@ -72,6 +83,7 @@ func ListenAndServe(addr string) error {
 	r.HandleFunc("/api/mailbox/{box}", postMessageHandler).Methods("POST")
 	r.HandleFunc("/api/posreport", postPositionHandler).Methods("POST")
 	r.HandleFunc("/api/status", statusHandler).Methods("GET")
+	r.HandleFunc("/api/gpsdPosition", positionHandler).Methods("GET")
```
[the remark this answers is not in the document]
Comment (blockmurder) [reply http.go]: Agreed. Changed the variable/function names in index.js as well.

File: http.go [hunk 4]
```diff
@@ -340,6 +352,40 @@ func getStatus() Status {
 
 func statusHandler(w http.ResponseWriter, req *http.Request) { json.NewEncoder(w).Encode(getStatus()) }
 
+func positionHandler(w http.ResponseWriter, req *http.Request) {
+	// Throw error if GPSd http endpoint is not enabled
```
[the remark this answers is not in the document]
Comment (blockmurder) [reply http.go]: Damn, missed this on. Old habits die hard... Thanks for the link!

**Comment by blockmurder** [comment]
@martinhpedersen  No worries, it's my pleasur! Seems like I missed hacking quite a bit during the last couple months ;-)

**Comment by martinhpedersen** [comment]
> No worries, it's my pleasur! Seems like I missed hacking quite a bit during the last couple months ;-)

I know the feeling 😁

**Comment by martinhpedersen** [comment]
Merged manually after rebase.

Funny story; Turns out that my GPS is affected by the rollover issue as well.

Thanks @blockmurder 😄 🛰

PS: Watch out for Hurricane Lorenzo!

**Comment by martinhpedersen** [comment]
FYI: See "The seven rules of a great Git commit message" @ https://chris.beams.io/posts/git-commit/.

Hope we can follow those rules in future commits :)

**Comment by blockmurder** [comment]
Sure, will do! Should have read https://github.com/la5nta/pat/blob/master/CONTRIBUTING.md more carefully...

**Pull request closed without merging**